feat: Pushing why you should use celery with rabbitmq (#1955)
section-engineering-education/engineering-education: articles/why-you-should-use-celery-with-rabbitmq/index.md

---
layout: engineering-education
status: publish
published: true
url: /engineering-education/why-you-should-use-celery-with-rabbitmq/
title: Why You Should Use Celery With RabbitMQ
description: This article will provide an overview of task queues & message brokers, why a user should choose celery with rabbitmq, and finally an example that shows how to use the two technologies.
author: edidiong-etuk
date: 2021-04-04T00:00:00-14:00
topics: [Languages]
excerpt_separator: <!--more-->
images:

  - url: /engineering-education/why-you-should-use-celery-with-rabbitmq/hero.png
    alt: Celery and RabbitMQ hero image
---

> In today's technology space, there is an increase in event-driven architectures as companies develop solutions that require asynchronous communication between their microservices. Ingesting these events faster in system architecture and processing them enables system architectures to be persistent, resilient and allows batch processing of data. Celery and RabbitMQ are some tools used in these architectures. Choosing the right combination of tools and viewing an example of these tools that goes beyond the ["hello world"](https://docs.celeryproject.org/en/stable/getting-started/brokers/rabbitmq.html) is what this article cover.

Reviewer: ```suggestion
In today's technology space, there is an increase in event-driven architectures as companies develop solutions that require asynchronous communication between their microservices. Ingesting these events faster in system architecture and processing them enables system architectures to be persistent, resilient and allows batch processing of data. Celery and RabbitMQ are some tools used in these architectures. Choosing the right combination of tools and viewing an example of these tools that go beyond the ["hello world"](https://docs.celeryproject.org/en/stable/getting-started/brokers/rabbitmq.html) is what this article cover.
```

### Outline

1. [What We Will Cover In This Article](#what-we-will-cover-in-this-article)
2. [Understanding Celery & RabbitMQ](#understanding-celery-&-rabbitmq)

>     - [What is a Task Queue](#what-is-a-queue)

Reviewer: ```suggestion   
    - [What is a Task Queue](#what-is-a-task-queue)
```

>     - [Examples of Task Queues](#examples-of-task-queues)

Reviewer: ```suggestion
    - [Examples of Task Queue Software](#examples-of-task-queue-software)
```

    - [What is a Message Broker](#what-is-a-message-broker)
    - [Examples of Message Brokers](#examples-of-message-brokers)
3. [Why We Should Choose The Celery & RabbitMQ Combo](#why-we-should-choose-the-celery-&-rabbitmq-combo)
4. [Setting up Celery with RabbitMQ](#setting-up-celery-with-rabbitmq)
    - [Installation & Configuration](#installation-&-configuration)
    - [Simple Use Case](#simple-use-case)
5. [Conclusion](#conclusion)

### What We Will Cover In This Article

> In this article, we’ll walk through what tasks queue are, see the examples of task queues (which includes celery), learn about message brokers. We will also look at why we need to use Celery and RabbitMQ while also seeing a basic tutorial that shows its installation and usage.

Reviewer: ```suggestion
In this article, we’ll walk through what task queues are, see the examples of task queues (which includes celery), learn about message brokers. We will also look at why we need to use Celery and RabbitMQ while also seeing a basic tutorial that shows its installation and usage.
```

### Understanding Celery & RabbitMQ

Before going into why we should use Celery & RabbitMQ, let us understand what they are and what similar tools are available. Celery is classified as a *Task Queue* and RabbitMQ is classified as a *Message Broker*. Let's dive deeper into what these are.

#### What is a Task Queue?

A task queue is a data structure maintained by a job scheduler containing jobs to run.

> Task queue software also manages background work that must be executed outside of the usual HTTP request-response cycle. They are designed for asynchronous work.

Reviewer: Please define what asynchronous work is, since a beginner might not be aware of async requests

To further explain their usefulness, let’s say we have a web application that uses artificial intelligence to enhance images.

> As the number of users increase, the time to enhance an image drastically increases, which leads to a significant delay while enhancing.

Reviewer: ```suggestion
As the number of users increases, the time to enhance an image drastically increases, which leads to a significant delay while enhancing.
```

> Hence, the need for a task queu/e software as it efficiently manages requests and ensures the application runs smoothly.

Reviewer: ```suggestion
Hence, the need for a task queue software as it efficiently manages requests and ensures the application runs smoothly.
```

#### Examples of Task Queue Software

- Celery
- Redis Queue
- Taskmaster
- Huey
- tasq

#### What is a Message Broker ?

> A message broker allows applications, systems and services to communicate and exchange information with each other. Now, with a task queue software in the background, the web app needs to know the status of the job in progress, plus any errors that might occur and the end result of the enhancement. A message broker dramatically eases this process since it’s built to make independent processes “talk to each other”

Reviewer: ```suggestion
A message broker allows applications, systems and services to communicate and exchange information with each other. Now, with a task queue software in the background, the web app needs to know the status of the job in progress, plus any errors that might occur and the end result of the enhancement. A message broker eases this process since it’s built to make independent processes “talk to each other”
```

#### Examples Of Message Brokers

- IBM MQ
- Beanstalk
- RabbitMQ
- Redis
- Gearman
- Solace

### Why We Should Choose The Celery & RabbitMQ Combo

> Having understood what task queues are, let's look at celery. Celery is an open-source task queue software written in python. It’s incredibly lightweight, supports multiple brokers (RabbitMQ, Redis and Amazon SQS) and it also integrates with multiple web frameworks, e.g. Django, etc. Celery's asynchronous task queue allows execution of tasks and its concurrency makes it used in several production systems. For example, Instagram uses Celery to scale thousands of tasks to millions.

Reviewer: ```suggestion
Having understood what task queues are, let's look at celery. Celery is an open-source task queue software written in python. It’s incredibly lightweight, supports multiple brokers (RabbitMQ, Redis and Amazon SQS) and also integrates with multiple web frameworks, e.g. Django, etc. Celery's asynchronous task queue allows the execution of tasks and its concurrency makes it used in several production systems. For example, Instagram uses Celery to scale thousands of tasks to millions.
```

However, task execution needs message brokers to work smoothly. Celery supports three message brokers as mentioned above. Although, for Amazon SQS, there is no support for remote monitoring.

> Redis as a message broker has its limitations, as doing replication is manual and requires additional work to implement it. Also, because Redis is an in-memory solution, if the machine runs out of memory when building queues up, there's a risk to lose tasks. Beanstalk also does not support replication.

Reviewer: ```suggestion
Redis, as a message broker has its limitations. Doing replication is manual and requires additional work to implement it. Also, because Redis is an in-memory solution, if the machine runs out of memory when building queues up, there's a risk to lose tasks. Beanstalk also does not support replication.
```

> Redis as a message broker has its limitations, as doing replication is manual and requires additional work to implement it. Also, because Redis is an in-memory solution, if the machine runs out of memory when building queues up, there's a risk to lose tasks. Beanstalk also does not support replication.

Reviewer: Please break this sentence down into smaller ones

RabbitMQ is the better choice as it guarantees message delivery, is fault tolerant, supports synchronous replication allows for SSL for establishing an encrypted connection, and it’s superb for real-time applications. In contrast, Redis has a problem with retaining data when a crash happens since it’s memory based and the SSL option is part of the paid version.

### Setting Up Celery With RabbitMQ

> For this setup, we will use an Ubuntu 18.04 machine to set up the celery app and the message queue (RabbitMQ). We'll configure the machine to work with celery and rabbitmq.

Reviewer: ```suggestion
For this setup, we will use an Ubuntu 18.04 machine to set up the celery app and the message queue (RabbitMQ). We'll configure the machine to work with celery and rabbitMQ.
```

Author: Hi, I intenionally used "rabbitmq" and not "RabbitMQ" because of SEO reasons..

Reviewer: Ah, alright then :)

#### Installation & Configuration

**Tools Needed:**

- virtualenv
- celery
- rabbitmq

We install and activate the virtual environment by entering the commands below on the terminal

```
sudo pip install virtualenv
mkdir my_project && cd my_project
virtualenv celery_project
source celery_project/bin/activate
```

> We install celery using pip. We don't use sudo as we’re installing celery to our virtual environment.

Reviewer: ```suggestion
We install celery using pip. We don't use sudo as we are installing celery to our virtual environment.
```

```
pip install celery
```

However, we also need to install rabbitmq on the system as it runs in the background.

```
sudo apt-get install rabbitmq-server
sudo rabbitmq-server -detached
```

The *`-detached`* option allows us to run *rabbitmq-server* in the background. Now, we could use defaults, but it’s always a good option to create a separate virtual host for our program.

```
sudo rabbitmqctl add_user myuser mypassword
sudo rabbitmqctl add_vhost myvhost
sudo rabbitmqctl set_permissions -p myvhost myuser ".*" ".*" ".*"
```

First  `".*"` *gives the user the ability to configure every entity, the* second *`".*"`* gives the *user write permissions on every entity*, and the third `".*"` gives the *user read permissions on every entity.*

> We install `dotenv` as we use environmental variables to protect sensitive information in our app.

Reviewer: Please explain briefly about environment variables

```
pip install python-dotenv
```

Then, we create and add our environmental variables in a `.env` file:

```
CELERY_BROKER_URL=amqp://myuser:mypassword@localhost/myvhost
CELERY_BACKEND_URL=db+sqlite:///test.db
```

As sqlite is used in our backend url, it needs to be installed alongside `sqlalchemy`. However, sqlite is available by default as it’s part of the standard libraries python uses.

```
pip install sqlalchemy
```

> Finally, we install flask as it’s our web framework to link the client with the server. requests is also installed to make api calls.

Reviewer: ```suggestion
Finally, we install `flask` as it’s our web framework to link the client with the server. The `requests` library is also installed to make api calls.
```

                  value="clearDogPics"
                  name="submit"
                >
                  Clear Photos
                </button>
              </div>
            </form>
          </div>
        </div>
        <div class="column">
          {% if link|length > 0 %}
          <ul>
            {% for links in link %}
            <li style="margin: 5px">
              <img src="{{links}}" />
            </li>
            {% endfor %}
          </ul>
          {% endif %}
        </div>
      </div>
    </section>
  </body>
</html>

```

Additional styling for the gallery done in our `main.css` file.

```css
ul {
  display: flex;
  flex-wrap: wrap;
}
li {
  height: 40vh;
  flex-grow: 1;
}
img {
  max-height: 100%;
  min-width: 100%;
  object-fit: cover;
  vertical-align: bottom;
}
li:last-child {
  flex-grow: 10;
}
@media (max-aspect-ratio: 1/1) {
  li {
    height: 30vh;
  }
}
@media (max-height: 480px) {
  li {
    height: 80vh;
  }
}
@media (max-aspect-ratio: 1/1) and (max-width: 480px) {
  ul {
    flex-direction: row;
  }
  li {
    height: auto;
    width: 100%;
  }
  img {
    width: 100%;
    max-height: 75vh;
    min-width: 0;
  }
}

```

> ![gendog1.png](/engineering-education/why-you-should-use-celery-with-rabbitmq/gendog1.png)

Reviewer: Please include a better alt for the image

Example: [Output Screen 1]

> ![gendog2.png](/engineering-education/why-you-should-use-celery-with-rabbitmq/gendog2.png)

Reviewer: Please provide a heading under the image, that describes what the image contains

```

First  `".*"` *gives the user the ability to configure every entity, the* second *`".*"`* gives the *user write permissions on every entity*, and the third `".*"` gives the *user read permissions on every entity.*

We install `dotenv` as we use environmental variables to protect sensitive information in our app. An environment variable is a variable that's defined outside of a program. This value can however be referenced at any point in time by a running program.

```bash
pip install python-dotenv
```

Then, we create and add our environmental variables in a `.env` file:

```bash
CELERY_BROKER_URL=amqp://myuser:mypassword@localhost/myvhost
CELERY_BACKEND_URL=db+sqlite:///test.db
```

As sqlite is used in our backend url, it needs to be installed alongside `sqlalchemy`. However, sqlite is available by default as it’s part of the standard libraries python uses.

```bash
pip install sqlalchemy
```

Finally, we install `flask` as it’s our web framework to link the client with the server. The `requests` library is also installed to make API calls.

```bash
pip install flask
pip install requests
```

### Simple Use Case

> To explain how task queues and message broker works, we’d take on an interesting project. It’s a Dog Pic Generator, so let's call it **GenDog**. It works by making an API call to **[dog.ceo](https://dog.ceo/dog-api/)** based on the breed selected by the user and the number of pictures selected. On refresh, the pictures of the dogs get updated on the page. The code for this example is at [this repository](https://github.com/edeediong/dog-generator-celery.git).

Author: @cyberShaw The link to the code is here.
